### PR TITLE
KNN SPMD python interfaces

### PR DESCRIPTION
**Pull request by ethanglaser**
# Description
Please include a summary of the change. For large or complex changes please include enough information to introduce your change and explain motivation for it.

Changes proposed in this pull request:
-
-
-

### CODE REVIEW

File: .ci/pipeline/ci.yml [hunk 1]
```diff
@@ -100,7 +100,7 @@ jobs:
     displayName: 'System info'
   - script: |
       conda update -y -q conda
-      if [ $(echo $(PYTHON_VERSION) | grep '3.7') ] || [ $(echo $(PYTHON_VERSION) | grep '3.11') ]; then export DPCPP_PACKAGE="dpcpp-cpp-rt>=2023.0.0"; else export DPCPP_PACKAGE="dpctl>=0.14"; fi
+      if [ $(echo $(PYTHON_VERSION) | grep '3.7') ] || [ $(echo $(PYTHON_VERSION) | grep '3.11') ]; then export DPCPP_PACKAGE="dpcpp-cpp-rt>=2023.1.0"; else export DPCPP_PACKAGE="dpctl>=0.14"; fi
```
Comment (samir-nasibli) [inline .ci/pipeline/ci.yml:103]: Why we doing it in `KNN SPMD python interfaces` PR?

File: examples/sklearnex/spmd_knn_bf_classification.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    size = comm.Get_size()
+
+    model_spmd, y_predict, y_test = run_example(rank, 100000, 100, 8, n_neighbors=20)
+
+    print("Brute Force Distributed kNN classification results:")
+    print("Ground truth (first 5 observations on rank {}):\n{}".format(rank, y_test[:5]))
+    print("Classification results (first 5 observations on rank {}):\n{}"
+          .format(rank, y_predict[:5]))
+    print("Accuracy for entire rank {} (256 classes): {}"
+          .format(rank, np.mean(np.equal(y_test, y_predict))))
```
Comment (samir-nasibli) [inline examples/sklearnex/spmd_knn_bf_classification.py:73]: Please avoid calling main. We agreed to simplify our sklearnex examples. See: https://github.com/intel/scikit-learn-intelex/blob/789d5478971b5bece05188eb35bae2b924b1dafa/examples/sklearnex/spmd_basic_statistics.py in #1205

File: .ci/pipeline/ci.yml (hunk 1 above)
Comment (ethanglaser) [reply .ci/pipeline/ci.yml]: I was just trying something, reverted now

File: examples/sklearnex/spmd_knn_bf_classification.py (hunk 2 above)
Comment (ethanglaser) [reply examples/sklearnex/spmd_knn_bf_classification.py:73]: Following Nikita's example for consistency (which was merged): https://github.com/intel/scikit-learn-intelex/blob/master/examples/sklearnex/spmd_linear_regression.py. Should I remove main?

Comment (samir-nasibli) [reply examples/sklearnex/spmd_knn_bf_classification.py:73]: Yeah. It is already updated on #1205. Try to follow. https://github.com/intel/scikit-learn-intelex/blob/master/examples/sklearnex/spmd_linear_regression.py also will be refactored and simplified. Thanks!

File: onedal/__init__.py [hunk 3]
```diff
@@ -50,3 +50,4 @@
 
     if _is_dpc_backend:
         __all__.append('spmd.linear_model')
+        __all__.append('spmd.neighbors')
```
Comment (samir-nasibli) [inline onedal/__init__.py:53]:
```suggestion
        __all__ += ['spmd.linear_model', 'spmd.neighbors']
```

File: onedal/neighbors/neighbors.py [hunk 4]
```diff
@@ -44,6 +44,12 @@
 
 
 class NeighborsCommonBase(metaclass=ABCMeta):
+    def _get_train_policy(self, queue, *data):
+        return _get_policy(queue, *data)
+
+    def _get_infer_policy(self, queue, *data):
+        return _get_policy(queue, *data)
```
Comment (samir-nasibli) [inline onedal/neighbors/neighbors.py:51]: Could you please comment why you are duplicating this? The same method with different names...

File: onedal/neighbors/neighbors.py [hunk 5]
```diff
@@ -568,14 +575,17 @@ def _onedal_predict(self, model, X, params, queue):
...
+        X = _convert_to_supported(policy, X)
         backend = _backend.neighbors.regression if gpu_device \
             else _backend.neighbors.search
 
         if hasattr(self, '_onedal_model'):
             model = self._onedal_model
         else:
             model = self._create_model(backend)
+        if gpu_device and "responses" not in params["result_option"]:
+            params["result_option"] += "|responses"
```
Comment (samir-nasibli) [inline onedal/neighbors/neighbors.py:588]: Move it to `_get_onedal_params` please.

File: onedal/spmd/neighbors/neighbors.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        raise NotImplementedError("predict_proba not supported in distributed mode.")
+
+
+class KNeighborsRegressor(NeighborsBaseSPMD, KNeighborsRegressor_Batch):
+    def predict(self, X, queue=None):
+        return self._predict_gpu(X, queue=queue)
+
+
+class NearestNeighbors(NeighborsBaseSPMD):
+    pass
```
Comment (samir-nasibli) [inline onedal/spmd/neighbors/neighbors.py:43]: If we don't support SPMD, there is no need adding it here

**Review by KulikovNikita — APPROVED** [review]
Everything looks fine for me. But I also propose to simplify examples a little bit

**Comment by samir-nasibli** [comment]
Please rebase your branch and address public CI test fails.

File: onedal/spmd/neighbors/neighbors.py (hunk 6 above)
Comment (ethanglaser) [reply onedal/spmd/neighbors/neighbors.py:43]: I think it should be left since it uses NeighborsBaseSPMD/NeighborsCommonBaseSPMD

File: onedal/neighbors/neighbors.py (hunk 5 above)
Comment (ethanglaser) [reply onedal/neighbors/neighbors.py:585]: It is dependent on gpu_device which depends on queue

Comment (samir-nasibli) [reply onedal/neighbors/neighbors.py:585]: > It is dependent on gpu_device which depends on queue

Function `_get_onedal_params` also have `queue` as an arg.

File: onedal/spmd/neighbors/neighbors.py (hunk 6 above)
Comment (samir-nasibli) [reply onedal/spmd/neighbors/neighbors.py:43]: We don't need `NeighborsBaseSPMD` as well here.

Comment (samir-nasibli) [reply onedal/spmd/neighbors/neighbors.py:43]: Remove it, please. Leave only NeighborsCommonBaseSPMD, the code where you really changing behavior. Remove `NearestNeighbors` from init file as well. If we don't support SPMD interface, no need to put this class here.

File: sklearnex/spmd/__init__.py [hunk 7]
```diff
@@ -14,4 +14,4 @@
 # limitations under the License.
 #===============================================================================
 
-__all__ = ['linear_model', 'basic_statistics']
+__all__ = ['linear_model', 'basic_statistics', 'neighbors']
```
Comment (samir-nasibli) [inline sklearnex/spmd/__init__.py:17]: alphabetic order.

File: sklearnex/spmd/neighbors/neighbors.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#===============================================================================
+
+from onedal.spmd.neighbors import (
+    KNeighborsClassifier,
+    KNeighborsRegressor,
+    NearestNeighbors
```
Comment (samir-nasibli) [inline sklearnex/spmd/neighbors/neighbors.py:20]: Remove it from here.

File: sklearnex/spmd/neighbors/__init__.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#===============================================================================
+
+from .neighbors import KNeighborsClassifier, KNeighborsRegressor, NearestNeighbors
+
+__all__ = ['KNeighborsClassifier', 'KNeighborsRegressor', 'NearestNeighbors']
```
Comment (samir-nasibli) [inline sklearnex/spmd/neighbors/__init__.py:19]: Remove please algos from the list if we don't support SPMD interface.

File: onedal/spmd/neighbors/neighbors.py (hunk 6 above)
Comment (ethanglaser) [reply onedal/spmd/neighbors/neighbors.py:43]: We do support it. It's the same circumstance as in [linear regression](https://github.com/intel/scikit-learn-intelex/blob/master/onedal/spmd/linear_model/linear_model.py#L27). No extensions of the class itself are introduced but there are extensions to classes it inherits from, so I believe it is necessary here.

Comment (samir-nasibli) [reply onedal/spmd/neighbors/neighbors.py:43]: There is no need for inheriting `NeighborsBaseSPMD`. This are not public. there is no need for such `NeighborsBaseSPMD`

File: onedal/spmd/neighbors/neighbors.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
```
Comment (samir-nasibli) [inline onedal/spmd/neighbors/neighbors.py:43]:
```suggestion
class NeighborsCommonBaseSPMD(ABC):
    def _get_policy(self, queue, *data):
        return _get_spmd_policy(queue)


class KNeighborsClassifier(NeighborsCommonBaseSPMD, KNeighborsClassifier_Batch):
    def predict_proba(self, X, queue=None):
        raise NotImplementedError("predict_proba not supported in distributed mode.")


class KNeighborsRegressor(NeighborsCommonBaseSPMD, KNeighborsRegressor_Batch):
    def predict(self, X, queue=None):
        return self._predict_gpu(X, queue=queue)


class NearestNeighbors(NeighborsCommonBaseSPMD):
    pass
```

File: examples/sklearnex/spmd_knn_bf_classification.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+model_spmd.fit(X_train, y_train, queue=q)
+
+y_predict = model_spmd.predict(X_test, queue=q)
+
+print("Brute Force Distributed kNN classification results:")
+print("Ground truth (first 5 observations on rank {}):\n{}".format(rank, y_test[:5]))
+print("Classification results (first 5 observations on rank {}):\n{}"
+      .format(rank, y_predict[:5]))
+print("Accuracy for entire rank {} (256 classes): {}"
+      .format(rank, np.mean(np.equal(y_test, y_predict))))
```
Comment (Alexsandruss) [inline examples/sklearnex/spmd_knn_bf_classification.py:64]: Use sklearn's accuracy score instead or even better classification score from sklearn.metrics

```suggestion
      .format(rank, np.mean(np.equal(y_test, y_predict))))
```

File: examples/sklearnex/spmd_knn_bf_regression.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+comm = MPI.COMM_WORLD
+rank = comm.Get_rank()
+size = comm.Get_size()
+
+if size < 2:
+    warn("This example was intentionally "
+         "designed to run in distributed mode only", RuntimeWarning)
+
+q = dpctl.SyclQueue("gpu")
```
Comment (Alexsandruss) [inline examples/sklearnex/spmd_knn_bf_regression.py:46]: GPU unavailability should be handled. For example, use exit with warning

File: onedal/neighbors/neighbors.py [hunk 13]
```diff
@@ -549,7 +552,8 @@ def _onedal_fit(self, X, y, queue):
 
             return train_alg(**params).compute(X, y).model
 
-        policy = _get_policy(queue, X, y)
+        policy = self._get_policy(queue, X, y)
+        X, y = _convert_to_supported(policy, X, y)
```
Comment (Alexsandruss) [inline onedal/neighbors/neighbors.py:556]: Is there any performance measurement to prove that this code didn't affect performance especially for small data shapes cases?

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: onedal/neighbors/neighbors.py (hunk 5 above)
Comment (ethanglaser) [reply onedal/neighbors/neighbors.py:585]: It does not currently have queue as an arg. I can add it as an optional arg and move lines into function or I can leave it as is to keep consistency with _get_onedal_params across all algos.

Comment (ethanglaser) [reply onedal/neighbors/neighbors.py:585]: But in current master there is a similar param conditional outside of _get_onedal_params which is part of the reason I configure it the way it is: https://github.com/intel/scikit-learn-intelex/blob/master/onedal/neighbors/neighbors.py#L430-L431

Comment (samir-nasibli) [reply onedal/neighbors/neighbors.py:585]: This is internal use func, there is no need saving consistency for these methods. It is not a public. So it could be and should extended in this case. Since it is on master you can keep it as is.
BTW kNN master is not good enough for the good code reference.

Comment (samir-nasibli) [reply onedal/neighbors/neighbors.py:585]: And also I didn't get why we adding condition for GPU device?

File: examples/sklearnex/spmd_knn_bf_regression.py (hunk 12 above)
Comment (ethanglaser) [reply examples/sklearnex/spmd_knn_bf_regression.py]: How does this seem - https://github.com/ethanglaser/scikit-learn-intelex/blob/dev/eglaser-spmd-knn/examples/sklearnex/spmd_knn_bf_regression.py#L46-L49

File: onedal/neighbors/neighbors.py (hunk 5 above)
Comment (ethanglaser) [reply onedal/neighbors/neighbors.py]: The addition was only necessary for distributed version so I think the latest commit most cleanly implements the logic and addresses your comments - https://github.com/ethanglaser/scikit-learn-intelex/blob/dev/eglaser-spmd-knn/onedal/spmd/neighbors/neighbors.py#L37-L41

**Comment by ethanglaser** [comment]
Thank you for your thorough reviews. Addressed most comments and finally external CI is green (was related to onedal version >= 20230100).

File: onedal/neighbors/neighbors.py (hunk 13 above)
Comment (ethanglaser) [reply onedal/neighbors/neighbors.py:556]: The inclusion or exclusion of this line does not impact performance (runtime, results) for examples of various sizes from a few basic experiments I have tried. @KulikovNikita have you observed improvements?

File: examples/sklearnex/spmd_knn_bf_regression.py (hunk 12 above)
Comment (Alexsandruss) [reply examples/sklearnex/spmd_knn_bf_regression.py]: I meant it should have better explanation why example can't be run. dpctl itself produces error saying 'device is unavailable'. @KulikovNikita, what do think about adding warning `GPU device is unavailable. Currently, SPMD execution mode is implemented only for this device type.`

File: setup.py [hunk 14]
```diff
@@ -475,9 +475,11 @@ def run(self):
           ] if ONEDAL_VERSION >= 20230100 else []
          ) + (
         ['onedal.spmd',
-         'onedal.spmd.linear_model',
-         'onedal.spmd.basic_statistics'
-         ] if build_distribute else [])),
+         'onedal.spmd.basic_statistics',
+         'onedal.spmd.linear_model'
+         ] + (['onedal.spmd.neighbors']
```
Comment (jmhatre-work) [inline setup.py:480]: ['onedal.spmd',
         'onedal.spmd.linear_model',
         'onedal.spmd.basic_statistics',
        'onedal.spmd.neighbors'
         ] instead of  +(['onedal.spmd.neighbors'])

Comment (ethanglaser) [reply setup.py:480]: I think the formatting makes the conditionals confusing, what the existing logic amounts to is:
if build_distributed:
  append(onedal.spmd, onedal.spmd.linear_model, onedal.spmd.basic_statistics]
  if onedal_version >= 20230100:
    append(onedal.spmd.neighbors)

File: examples/sklearnex/spmd_knn_bf_classification.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
+#===============================================================================
```
Comment (samir-nasibli) [inline examples/sklearnex/spmd_knn_bf_classification.py:1]: Please rename your examples like `examples/skleaernex/<algo_name>_spmd.py`.

**Comment by samir-nasibli** [comment]
Hi just add GPU device and librariry requirements for your examples here https://github.com/intel/scikit-learn-intelex/blob/master/tests/run_examples.py#L127:L156:
```
...
req_device['knn_bf_classification_spmd.py'] = ["gpu"]
req_device['knn_bf_regression_spmd.py'] = ["gpu"]

...

req_library['knn_bf_classification_spmd.py'] = ['catboost']
req_library['knn_bf_regression_spmd.py''] = ['dpctl', 'mpi4py']
```
And also rename your examples and make `spmd` as suffix in the file names.

File: onedal/spmd/neighbors/neighbors.py [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+
+class KNeighborsClassifier(NeighborsCommonBaseSPMD, KNeighborsClassifier_Batch):
+    def predict_proba(self, X, queue=None):
+        raise NotImplementedError("predict_proba not supported in distributed mode.")
+
+
+class KNeighborsRegressor(NeighborsCommonBaseSPMD, KNeighborsRegressor_Batch):
+    def predict(self, X, queue=None):
+        return self._predict_gpu(X, queue=queue)
```
Comment (samir-nasibli) [inline onedal/spmd/neighbors/neighbors.py:35]: What if it is not GPU queue?

**Review by Alexsandruss — APPROVED** [review]

**Comment by ethanglaser** [comment]
Should these interfaces be compatible with using dpctl tensors for X and y without queue argument? Because currently it is not. And it seems like other algos may not be also, not sure if this is needed at the moment.

**Comment by samir-nasibli** [comment]
> Should these interfaces be compatible with using dpctl tensors for X and y without queue argument? Because currently it is not. And it seems like other algos may not be also, not sure if this is needed at the moment.

This is good question and it is common for all SPMD iface algos. We are working on DCPTL adoption, so it will be aligned on  https://github.com/intel/scikit-learn-intelex/pull/1216.

This PR is done :)

**Review by samir-nasibli — APPROVED** [review]
